Address final review comments (#17)
OpenLiberty/draft-guide-microprofile-config-apis: README.adoc

// =================================================================================================
== What you'll learn

You'll learn how to improve flexibility and maintainability of your application's configurations by using https://openliberty.io/docs/latest/reference/feature/mpConfig-3.0.html[MicroProfile Config^]. 

> MicroProfile Config streamlines the configuration process for microservices by providing a flexible mechanism for application configuration. Application can use the https://openliberty.io/docs/latest/reference/javadoc/microprofile-5.0-javadoc.html?package=org/eclipse/microprofile/config/package-frame.html&class=org/eclipse/microprofile/config/package-summary.html[MicroProfile Config API^] to eliminate the need for repetitive injection statements and simplifies the handling of complex configurations with bulk-extraction of configuration properties, dynamic property expressions, and retrieval of multi-valued properties as lists. Additionally, the configuration source APIs allow for fine-grained control and access to the metadata and values of configuration properties, enabling greater flexibility and customization in managing application configurations.

Reviewer: ```suggestion
MicroProfile Config streamlines the configuration process for microservices by providing a flexible mechanism for application configuration. Applications can use the https://openliberty.io/docs/latest/reference/javadoc/microprofile-5.0-javadoc.html?package=org/eclipse/microprofile/config/package-frame.html&class=org/eclipse/microprofile/config/package-summary.html[MicroProfile Config API^] to eliminate the need for repetitive injection statements and simplify the handling of complex configurations with bulk-extraction of configuration properties, dynamic property expressions, and retrieval of multi-valued properties as lists. Additionally, the configuration source APIs allow for fine-grained control and access to the metadata and values of configuration properties, enabling greater flexibility and customization in managing application configurations.
```

// =================================================================================================
== Defining a property with expressions

> In MicroProfile Config, property expressions provide a way to set and expand variables in property values. To use property expressions, define a property with a `${expression}` syntax, where `expression` is the name of another configuration property. When the property is loaded, the `${expression}` syntax is replaced with the value of the referenced property.

Reviewer: I think the value of property expressions is that you can make the value of one property depend on the value of another property- can you think of a concise example where this would be beneficial? perhaps when you need to specify a port value or similar? Doesnt need to be much but could go in between the first and second sentence in this paragraph.

Author: Edited in https://github.com/OpenLiberty/draft-guide-microprofile-config-apis/pull/17/commits/cdee87d7a502e51b1c3328cfe2e7972aaf4d5a3c

mvn liberty:dev
```

After you see the following message, your application server in dev mode is ready:

[role="no_copy"]
----
**************************************************************
*    Liberty is running in dev mode.
----

Dev mode holds your command-line session to listen for file changes. Open another command-line session to continue, or open the project in your editor.

// file 0
query/pom.xml
[source, XML, linenums, role='code_column']
----
include::finish/query/pom.xml[]
----

// file 1
query/server.xml
[source, XML, linenums, role='code_column']
----
include::finish/query/src/main/liberty/config/server.xml[]
----

The MicroProfile Config API is included in the MicroProfile dependency that is specified in your [hotspot file=0]`query/pom.xml` file. Look for the dependency with the [hotspot=mpconfig file=0]`microprofile` artifact ID. This dependency provides a library that allows you to use the MicroProfile Config API to externalize configurations for your microservices. The [hotspot=mpconfig file=1]`mpConfig` feature is also enabled in the [hotspot file=1]`query/src/main/liberty/config/server.xml` file.

> In MicroProfile Config, property expressions provide a way to set and expand variables in property values. To use property expressions, define a property with a `${expression}` syntax, where `expression` is the name of another configuration property. When the property is loaded, the `${expression}` syntax is replaced with the value of the referenced property.

Reviewer: i think this paragraph and the one at line 151 could be moved to the beginning of this section- that would set the context before the reader starts the config task

Author: Resolved in https://github.com/OpenLiberty/draft-guide-microprofile-config-apis/pull/17/commits/3d195954e4f0bf9f6a80e3696735774b6102941d

include::finish/query/src/main/resources/META-INF/microprofile-config.properties[]
----

Define a property called [hotspot=userPassword file=2]`system.userPassword` that references two other properties, [hotspot=user file=2]`system.user` and [hotspot=password file=2]`system.password`. 

The `${system.user}` and `${system.password}` expressions are replaced with their respective values when the [hotspot=userPassword file=2]`system.userPassword` property is loaded, resulting in `alice:alicepwd`.

> To use this configuration property,

Reviewer: ```suggestion
To use this configuration property, add it to the `SystemResource` class.
```

Note that only registered users can access the system microservice. To authenticate a user, a [hotspot=basicRegistry file=4]`basicRegistry` element is configured in the `system` server configuration file that is at [hotspot file=4]`system/src/main/liberty/config/server.xml`. This `basicRegistry` element is a simple case for learning purposes, and for more information about different user registries, see the https://openliberty.io/docs/latest/user-registries-application-security.html[User registries documentation^].

Because you are running the `query` service in dev mode, the changes that you made were automatically picked up. You're now ready to check out your application in your browser.

Point your browser to the http://localhost:9085/query/systems/localhost URL. You can see the current OS name and Java version in JSON format.

// ===========================================================================================

> == Injecting property as a list

Reviewer: what is the benefit of doing this- or an example case where it would be useful?

Author: Edited in https://github.com/OpenLiberty/draft-guide-microprofile-config-apis/pull/17/commits/324a3685c77af4f7dfe3051751832cbd805d968f

> Configuration values are purely Strings. MicroProfile Config API has built-in converters that automatically convert configured Strings into target types. When injecting a multi-valued property, MicroProfile Config API assumes that the property value is a comma-separated list of items. It then splits the value into individual items and returns them as a List of Strings.

Reviewer: ```suggestion
Configuration values are just Strings. MicroProfile Config API has built-in converters that automatically convert configured Strings into target types. When injecting a multi-valued property, MicroProfile Config API assumes that the property value is a comma-separated list of items. It then splits the value into individual items and returns them as a List of Strings.
```

include::finish/query/src/main/java/io/openliberty/guides/query/SystemResource.java[]
----

Inject the `system.properties` property by using the [hotspot=systemPropertiesInject file=1]`@Inject` and [hotspot=systemPropertiesProperty file=1]`@ConfigProperty` annotations with a variable type of [hotspot=listStringType file=1]`List<String>`.

The [hotspot=listStringType file=1]`systemProperties` variable retrieves the list of system properties from the configuration and passes it to the [hotspot=getSystemPropertiesMethod file=1]`getSystemProperties()` method.

The [hotspot=getSystemPropertiesMethod file=1]`getSystemProperties()` then iterates through this list of system properties by using the [hotspot=systemProperties file=1]`systemProperties` variable to retrieve its value from the `system` service.

Point your browser to the http://localhost:9085/query/systems/localhost URL. You can see the list of system properties that are retrieved from the `system.properties` property in JSON format.

// ===========================================================================================
== Using configuration source APIs

> MicroProfile Config provides https://openliberty.io/docs/latest/reference/javadoc/microprofile-5.0-javadoc.html?package=org/eclipse/microprofile/config/package-frame.html&class=org/eclipse/microprofile/config/package-summary.html[configuration source APIs^] for obtaining detail information about your application's configuration. These APIs can be used to retrieve various useful information about a single specified configuration property by using the `ConfigValue` API class, or to lookup the metadata of all underlying configurations of your microservice by using the `Config` API class.

Reviewer: same here- whats the benefit?

Author: Edited in https://github.com/OpenLiberty/draft-guide-microprofile-config-apis/pull/17/commits/ae426b2edaa323381d31a623506df60e6220e304

include::finish/query/src/main/java/io/openliberty/guides/query/SystemResource.java[]
----

Inject the `system.properties` property by using the [hotspot=systemPropertiesInject file=1]`@Inject` and [hotspot=systemPropertiesProperty file=1]`@ConfigProperty` annotations with a variable type of [hotspot=listStringType file=1]`List<String>`.

The [hotspot=listStringType file=1]`systemProperties` variable retrieves the list of system properties from the configuration and passes it to the [hotspot=getSystemPropertiesMethod file=1]`getSystemProperties()` method.

The [hotspot=getSystemPropertiesMethod file=1]`getSystemProperties()` then iterates through this list of system properties by using the [hotspot=systemProperties file=1]`systemProperties` variable to retrieve its value from the `system` service.

Point your browser to the http://localhost:9085/query/systems/localhost URL. You can see the list of system properties that are retrieved from the `system.properties` property in JSON format.

// ===========================================================================================
== Using configuration source APIs

> MicroProfile Config provides https://openliberty.io/docs/latest/reference/javadoc/microprofile-5.0-javadoc.html?package=org/eclipse/microprofile/config/package-frame.html&class=org/eclipse/microprofile/config/package-summary.html[configuration source APIs^] for obtaining detail information about your application's configuration. These APIs can be used to retrieve various useful information about a single specified configuration property by using the `ConfigValue` API class, or to lookup the metadata of all underlying configurations of your microservice by using the `Config` API class.

Reviewer: ```suggestion
MicroProfile Config provides https://openliberty.io/docs/latest/reference/javadoc/microprofile-5.0-javadoc.html?package=org/eclipse/microprofile/config/package-frame.html&class=org/eclipse/microprofile/config/package-summary.html[configuration source APIs^] for obtaining detailed information about your application configuration. These APIs can be used to retrieve various useful information about a single specified configuration property by using the `ConfigValue` API class, or to look up the metadata of all underlying configurations of your microservice by using the `Config` API class.
```

> The configuration source APIs offer a powerful toolset for managing and utilizing your microservice's configuration in a flexible and efficient manner. They enable you to add configurations from various sources, such as the `server.xml`, `server.env`, and `bootstrap.properties` files, or ConfigSources for different environments, such as `development`, `testing`, and `production`. With these APIs, you can easily determine which sources and properties are loaded in your current environment and make changes accordingly. Additionally, this approach can be useful for debugging and troubleshooting purposes.

Reviewer: ```suggestion
The configuration source APIs offer a powerful toolset for managing and utilizing your microservice configuration in a flexible and efficient manner. They enable you to add configurations from various sources, such as the `server.xml`, `server.env`, and `bootstrap.properties` files, or ConfigSources for different environments, such as `development`, `testing`, and `production`. With these APIs, you can easily determine which sources and properties are loaded in your current environment and make changes accordingly. These APIs can also be useful for debugging and troubleshooting purposes.
```

Then, inject the [hotspot=config file=1]`Config` metadata object, and add the [hotspot=getConfigSourcesClassMethod file=1]`getConfigSources()` and the [hotspot=getConfigProperties file=1]`getConfigProperties()` class methods.

The [hotspot=getConfigSources file=1]`Config.getConfigSources()` method in the Config API returns all the registered ConfigSources for the current service. These ConfigSources contain the configuration properties that are available to the service. 

The [hotspot=getPropertyNames file=1]`Config.getPropertyNames()` method in the Config API finds all configuration property names by searching through all the registered ConfigSources.

Now, check out the following endpoints that you just created:

* The http://localhost:9085/query/config/contact URL returns the source name, source ordinal, and value of the `query.contactEmail` property.

* The http://localhost:9085/query/config URL returns all registered configuration sources and properties of the running `query` service.

// ===========================================================================================
== Injecting configuration as a bulk through @ConfigProperties annotation

> MicroProfile Config provides a practical way to aggregate related configuration properties into a single CDI bean by using the `@ConfigProperties` annotation. This annotation is applied to a plain old Java object (POJO) that serves as a holder for configuration properties, making it easier to retrieve. By injecting the annotated POJO as a CDI bean into a separate POJO, a cleaner and more organized approach to managing configuration properties is achieved.

Reviewer: ```suggestion
MicroProfile Config provides a practical way to aggregate related configuration properties into a single CDI bean by using the `@ConfigProperties` annotation. This annotation is applied to a plain old Java object (POJO) that serves as a holder for configuration properties, making it easier to retrieve. Injecting the annotated POJO as a CDI bean into a separate POJO provides a cleaner and more organized approach to managing configuration properties.
```
